File: src/bmsx/emulator/instruction_format.ts
export const INSTRUCTION_BYTES = 4;

export const MAX_OP_BITS = 6;
export const MAX_OPERAND_BITS = 6;
export const MAX_BX_BITS = 12;

export const EXT_A_BITS = 2;
export const EXT_B_BITS = 3;
export const EXT_C_BITS = 3;
export const EXT_BX_BITS = 8;

export const MAX_LOW_OPERAND = (1 << MAX_OPERAND_BITS) - 1;
export const MAX_LOW_BX = (1 << MAX_BX_BITS) - 1;
export const MAX_WIDE = (1 << MAX_OPERAND_BITS) - 1;

export const MAX_BASE_OPERAND_A = (1 << (MAX_OPERAND_BITS + EXT_A_BITS)) - 1;
export const MAX_BASE_OPERAND_BC = (1 << (MAX_OPERAND_BITS + EXT_B_BITS)) - 1;
export const MAX_BASE_BX = (1 << (MAX_BX_BITS + EXT_BX_BITS)) - 1;

export const MAX_EXT_REGISTER_A = (MAX_WIDE << (MAX_OPERAND_BITS + EXT_A_BITS)) | MAX_BASE_OPERAND_A;
export const MAX_EXT_REGISTER_BC = (MAX_WIDE << (MAX_OPERAND_BITS + EXT_B_BITS)) | MAX_BASE_OPERAND_BC;
export const MAX_EXT_REGISTER = MAX_EXT_REGISTER_BC;
export const MAX_EXT_CONST = (1 << (MAX_OPERAND_BITS + EXT_B_BITS - 1)) - 1;
export const MAX_EXT_BX = (MAX_WIDE << (MAX_BX_BITS + EXT_BX_BITS)) | MAX_BASE_BX;


export function packInstructionWord(op: number, a: number, b: number, c: number, ext: number = 0): number {
	return ((ext & 0xff) << 24)
		| ((op & 0x3f) << 18)
		| ((a & 0x3f) << 12)
		| ((b & 0x3f) << 6)
		| (c & 0x3f);
}

export function writeInstruction(code: Uint8Array, index: number, op: number, a: number, b: number, c: number, ext: number = 0): void {
	const word = packInstructionWord(op, a, b, c, ext);
	const offset = index * INSTRUCTION_BYTES;
	code[offset] = (word >>> 24) & 0xff;
	code[offset + 1] = (word >>> 16) & 0xff;
	code[offset + 2] = (word >>> 8) & 0xff;
	code[offset + 3] = word & 0xff;
}

export function readInstructionWord(code: Uint8Array, index: number): number {
	const offset = index * INSTRUCTION_BYTES;
	return (code[offset] << 24)
		| (code[offset + 1] << 16)
		| (code[offset + 2] << 8)
		| code[offset + 3];
}

File: src/bmsx_cpp/machine/cpu/instruction_format.cpp
#include "machine/cpu/instruction_format.h"
#include <cstddef>

namespace bmsx {

int signExtend(uint32_t value, int bits) {
	const int shift = 32 - bits;
	return static_cast<int>(value << shift) >> shift;
}

uint32_t packInstructionWord(uint8_t op, uint8_t a, uint8_t b, uint8_t c, uint8_t ext) {
	return (static_cast<uint32_t>(ext) << 24)
		| (static_cast<uint32_t>(op & 0x3f) << 18)
		| (static_cast<uint32_t>(a & 0x3f) << 12)
		| (static_cast<uint32_t>(b & 0x3f) << 6)
		| static_cast<uint32_t>(c & 0x3f);
}

void writeInstruction(std::vector<uint8_t>& code, int index, uint8_t op, uint8_t a, uint8_t b, uint8_t c, uint8_t ext) {
	const uint32_t word = (static_cast<uint32_t>(ext) << 24)
		| (static_cast<uint32_t>(op & 0x3f) << 18)
		| (static_cast<uint32_t>(a & 0x3f) << 12)
		| (static_cast<uint32_t>(b & 0x3f) << 6)
		| static_cast<uint32_t>(c & 0x3f);
	writeInstructionWord(code, index, word);
}

void writeInstructionWord(std::vector<uint8_t>& code, int index, uint32_t word) {
	const size_t offset = static_cast<size_t>(index) * INSTRUCTION_BYTES;
	code[offset] = static_cast<uint8_t>((word >> 24) & 0xff);
	code[offset + 1] = static_cast<uint8_t>((word >> 16) & 0xff);
	code[offset + 2] = static_cast<uint8_t>((word >> 8) & 0xff);
	code[offset + 3] = static_cast<uint8_t>(word & 0xff);
}

uint32_t readInstructionWord(const std::vector<uint8_t>& code, int index) {
	const size_t offset = static_cast<size_t>(index) * INSTRUCTION_BYTES;
	return (static_cast<uint32_t>(code[offset]) << 24)
		| (static_cast<uint32_t>(code[offset + 1]) << 16)
		| (static_cast<uint32_t>(code[offset + 2]) << 8)
		| static_cast<uint32_t>(code[offset + 3]);
}

} // namespace bmsx
